### ISSUE TITLE
myEYPD: Let learners add professional development hours (learner)

### ISSUE THREAD

**Issue by alex-418** [BCcampus/eypd#264]
@kathreenriel  Adding this here for your validation. 

**Request:** Consider allowing Learners to enter their own training and hours manually in their myEYPD to more accurately represent their total training hours. 

**Use case:** A learner previously attended an event (before the "Plan to Attend" feature was made available) so they have no way of adding those event hours. Or they attended, and forgot to hit the "plan to attend button. Or the event they attended wasn't posted on the EYPD website.

I was sent this excel file by a learner (Crystal Janes -  Director of Programs, Westcoast Child Care) who has to manually track her events not on EYPD website, it would be great if we could provide a way for them to not have to do this manually: 
![tracking](https://user-images.githubusercontent.com/17072191/31748538-c4b5af6a-b428-11e7-80e6-e6615442bd56.png)

**Comment by mandily-p** [comment]
In addition to just tracking hours in a table, It may also be worth investigating the work needed to allow users to upload the certificates that correspond with the events into the system. Even if it's just as a planning step for a future sprint.

**Comment by mandily-p** [comment]
Here's a first draft mock. @kathreenriel What fields so you think we need to collect? I see the spreadsheet has location and facilitator, but the person who sent it to you didn't fill those in. I was thinking the building location and city would be helpful. What do you think?

The user would arrive at this modal by the link you can see peeking through underneath - A "Add an event I've attended." I'm wondering if it might not be better to say "Record an event I've attended"?

In the learning domains, I was thinking we could use a jquery replacement for select boxes called Select2: https://select2.org/tagging#tagging-with-multi-value-select-boxes

The date box would have a calendar picker

the hours box would be client side validated to only allow numbers up to one decimal point.

I also like the idea of giving the user the option to add multiple at a time instead of closing the modal and then having to open it again. That's what the secondary button style is in the mock.

Thoughts?

![add new event](https://user-images.githubusercontent.com/20425828/44566248-4baa2000-a721-11e8-9a42-dad5843d8c0e.png)

**Comment by alex-418** [comment]
@mandily-p nice, thanks for this I was thinking something similar. Here's some input for your consideration. 

**Date picker:** 
Should be straight forward to implement (we're already using one in the cert hours) 

**Query select box replacement:** 
I like this, but unsure of the value add without some sort of validation from users/research? And we should think about doing this consistently across the site, along with making sure it works on different browsers/mobile etc. 

**Add another:** 
This could be straight-forward, or a time sink to implement. Depends on the desired functionality of the button.

**Comment by mandily-p** [comment]
**More notes on Select2**
I've used this in multiple projects before. It's always tested well with users and is a fairly low cost way to get a lot of options into a small space and have it still be manageable. The other option here would be listing out all of the categories in checkboxes.

**More notes on the Add another**
I envisioned showing the user a message that the event was added and clearing out the form for them to enter the next item.
![add new event - multiple](https://user-images.githubusercontent.com/20425828/44567262-83679680-a726-11e8-92f5-dd1c0496b8f0.png)

**Comment by alex-418** [comment]
**Select2:**
Great, thanks for adding those notes on Select2... definitely more elegant that our current solution: 
![select_list](https://user-images.githubusercontent.com/17072191/44567420-23252480-a727-11e8-80a6-efea0375cdf8.png)

**Add another**
Ok, great, so basically Submit the form in the background, add success message, and clear the fields. Shouldn't bee too difficult, will work on getting the modal working first and see where we're at.

**Comment by alex-418** [comment]
Update: I've created the modal and working on passing the data back, so I can in the background save the new event information. One thing that we can't avoid is a reload of the page because the other elements on the page require this to update the data. So i've made that happen when the "save" past events button is clicked. 

@kathreenriel @mandily-p  Please note: This won't update the donut, only the past events list. For donut to update we require the user to hit the "Calculate My Hours" button as well... not very ideal, so if that's a deal-breaker, I can spend some time to see what would need to be done to myEYPD for this to behave more like a single page application. (although, it would be time consuming to execute on)

**Comment by mandily-p** [comment]
How much effort would be involved in adding multiple states to the save button so the user would be able to tell if they needed to save the page to see updates vs "clicking just in case"? It's not the front end work I'm wondering about here, but more about whether it's "easy" to check for stale data on the page? Or maybe it's a check to see if the add function has been called but the save button hasn't yet?

I dunno. Any ideas that would be less costly than auto updating? (although I'm still going to add a future enhancement ticket for auto updating :) )

**Comment by alex-418** [comment]
I'll see what I can do. But working outside the current limitations will be significant effort. There's no easy way to accomplish this. I also feel like there might be value in defining the exact problem we're trying to solve in that new issue you created. Especially if there will continue to be changes to this. The time spent in piecing it together like this, might be better spent solving the root problem. Anyhow, will update with progress once I've made some significant headway.

**Comment by mandily-p** [comment]
I'm really just trying to minimize user interaction for them to get the information they want. But I'd say this is low priority because it's more of a polish item than a necessity. Let's hold off on any extra work until further down the road. There's bigger fish to fry at the moment :)

**Comment by alex-418** [comment]
Yes, totally agree. But from my end as well... it's pretty frustrating to not be able to quickly deliver on that. Hate to put so many steps/decisions on the user to get what they need.

**Comment by kathreenriel** [comment]
Thanks @alex-418 @mandily-p for working through these options/ideas
Here's some feedback. 
1. Change "Learning Domains" to "Categories"
2. Re:  "multiple states save" @alex-418 this is significant progress on this issue we can make refinements at a later date. 
3. Question about jquery replacement  how do users know the categories without prompts or a list to choose from?  I would like to keep the categories consistent between what is posted by Organizers and what is added by Learners

**Comment by alex-418** [comment]
Progress in https://github.com/BCcampus/eypd/pull/544

**Comment by bdolor** [comment]
@kathreenriel - With a data creation mechanism in place, usually what follows is the desire for *updating/editing* that data, and/or *deleting* that data. As it currently is, once it's recorded there's no way for anyone to do anything but look at it. At first blush, I'm thinking this needs to be scoped out a bit more.

**Comment by mandily-p** [comment]
Blocked by #598 - Add delete and edit functionality into the past events table.

**Comment by kathreenriel** [comment]
This is a high priority function for future development of the EYPD website

**Comment by dshaykewich** [comment]
Won't fix.
